### PR TITLE
mempool: keep CPFP'd transactions when loading from mempool.dat

### PR DESCRIPTION
**Pull request by glozow**
Part of v3 and package relay (see #27463).

**Problem**
When loading mempool.dat, we apply -minrelaytxfee and mempool min feerate on each transaction, meaning we'll reject transactions that may be CPFP'd by later transactions mempool.dat. Even without package relay, we can run into this problem if we are shrinking -maxmempool or raising -minrelaytxfee on a restart.

**Solution**
When loading mempool.dat, use `bypass_limits=true` and then call `TrimToSize()` at the very end.

Advantages:
- We definitely keep the "highest descendant score" transactions if mempool min feerate rises.
- It's extremely simple implementation-wise.
- It's simple to keep track of what made it in and what didn't.

Disadvantages:
- If mempool.dat is very large, we can exceed maxmempool by quite a bit.
- This won't be sufficient for ephemeral anchors (unless `bypass_limits` allows not-yet-spent anchors).

### CODE REVIEW

**Comment by ajtowns** [comment]
Maybe move most of the text in the PR description into a comment? "Review this first" and "Alternatives considered" stuff is useful here, but doesn't seem useful for the merge commit message?

**Comment by glozow** [comment]
**Alternatives Considered**

Feel free to suggest more alternatives, but note potential problems.

(1) Call TrimToSize() at smart intervals, e.g. when a timestamp changes (since packages are submitted with the same timestamp), or every ~25 transactions, or when we hit a certain threshold above maxmempool.
(2) When a transaction fails for fee-related reasons, try again with bypass_limits=true, and then schedule a flush for later.
(3) Create a new mempool.dat format, which contains a list of packages instead of transactions (package can have 1 transaction).
(4) When loading, load into a buffer and use AncestorPackage "packageifier" (in https://github.com/bitcoin/bitcoin/pull/26711) to group them into packages and submit them together.

Their disadvantages:

-    (1, 2) The implementation is a bit hacky. Some strategies may fail.
-    (3, 4) The implementation is quite complex.
-    (1, 2) When shrinking maxmempool, we won't always pick the top transactions by descendant feerate. Since mempool min feerate depends on the feerate of the last package removed + incremental, we can end up rejecting something with feerate 1sat/vB higher than what we keep. Essentially, we have a bias for earlier transactions. This is weird behavior (and makes it quite annoying to write tests).
-    (1, 2) We need to add various bits of code to track things evicting each other to accurately report which transactions made it in and which didn't.
-    (1, 2) Doesn't work for ephemeral anchors (https://github.com/bitcoin/bitcoin/pull/26403). We could work around this by adding another flag to bypass "unspent anchor" checks but... also a bit ugly.

We can also pick one of the more complex solutions to implement later (I expect people are going to be interested in (3) since it's the most future-proof). This problem shouldn't be very common right now. For the most part, it doesn't really matter if the "bottom" of your mempool doesn't match others'.

**Comment by glozow** [comment]
> Maybe move most of the text in the PR description into a comment? "Review this first" and "Alternatives considered" stuff is useful here, but doesn't seem useful for the merge commit message?

Moved "alternatives considered." I prefer putting "review this first" at the top so reviewers see it as soon as possible. It will go away if/when I rebase, so definitely before any merging happens.

File: src/kernel/mempool_persist.cpp [hunk 1]
```diff
@@ -63,6 +63,7 @@ bool LoadMempool(CTxMemPool& pool, const fs::path& load_path, Chainstate& active
         }
         uint64_t num;
         file >> num;
+        LOCK2(cs_main, pool.cs);
```
Comment (dergoegge) [inline src/kernel/mempool_persist.cpp:66]: Afaict we currently import mempool txs from disk in the background and allow the node to make connections and sync the chain in parallel. Seems like this commit changes that by locking cs_main for the entire import duration. Not sure about the consequences.

Potentially relevant for #27460 as well.

**Review by dergoegge — COMMENTED** [review]
Approach ACK - importing all and then trimming seems like the least complex approach to me.

File: src/kernel/mempool_persist.cpp (hunk 1 above)
Comment (instagibbs) [reply src/kernel/mempool_persist.cpp:66]: To recap, this disallows a wallet(or other TrimToSize caller) from trimming a parent before the child is entered into the mempool?

With debug builds, this draws out the startup time of bitcoind by ~1.5 minutes with 300MB mempool. ~30 seconds on non-debug build.

**Review by instagibbs — COMMENTED** [review]
approach ACK

**Comment by glozow** [comment]
Closing this for now to (1) focus on more important functionality (2) defer the decision-making to when we have more information on usage and/or better eviction.
- The "trim stuff below minrelayfeerate" thing isn't very good... as established, we might evict things we shouldn't, etc.
- This isn't strictly necessary yet until package relay is deployed.
- It might be too early to decide on a solution. Perhaps with EA and actual package relay usage, we'll end up wanting to reformat mempool.dat anyway. Or maybe, with other improvements to mempool, we will have a simpler solution that achieves everything we want.

Still tracking this as a "todo" item in #27463, but will defer until later.

**Review by ajtowns — COMMENTED** [review]
Approach NACK: I don't think holding cs_main or the mempool lock for the full import period is okay.

I'm not sure this needs fixing at all at this point: it doesn't affect nodes that don't restart in between the tx being broadcast and mined; it doesn't affect nodes that don't have a full mempool (eg, if you raise the limit to 1GB or 2GB instead of 300MB); and it's not needed if the child tx is either rebroadcast or rbf'ed. In some sense, saving the mempool and reloading it on restart is just a nice-to-have, so having some txs be missing on restart is at worst annoying, not a fatal bug.

It would however be nice to fix before we start accepting below-minrelaytxfee txs as part of packages (eg 0-fee txs whose fees are paid by spending an ephemeral anchor output) -- in that case parents will be lost on restart even if the mempool is essentially empty, which would be annoying.

Here's a branch that takes an alternative approach of deferring `TrimToSize` until the mempool is loaded: https://github.com/ajtowns/bitcoin/commits/202305-defertrimtosize The idea there is to not consider it a DoS to (eg) load a 1GB mempool.dat in its entirety even if you have a 300MB limit, but to still protect against txs from peers pushing memory usage substantially above that limit.

That feels acceptable to me, though YMMV. Other approaches we could consider while remaining backwards compatible (ie without changing the mempool.dat format):

 * special case below-minrelaytxfee txs: when AcceptToMemoryPool detects `insufficient fee`, report that directly (not as a generic `MEMPOOL_POLICY` failure), and have `LoadMempool` keep track of all those txs so they can be included if a later tx considers them an ancestor. That means you may go over the maxmempool limit, but only by the size of the below-minrelaytxfee txs from your mempool.dat. If those are small (due to v3/ephemeral anchor rules?) anyway, that might be acceptable; you might also only need to keep `limitancestorcount` of them, if you write the ancestor that cpfps them to disk asap when saving.
 * implement cluster mempool first and export mempool.dat in chunks from highest to lowest fee;  when importing, assume that the mempool contains chunks from highest score to lowest, build up a chunk to import; once it's over the mempool minfee, import the txs immediately; if it's over 1MvB and still under the mempool minfee, drop it and stop attempting to import txs from this mempool.dat entirely?

File: src/kernel/mempool_persist.cpp (hunk 1 above)
Comment (glozow) [reply src/kernel/mempool_persist.cpp:66]: To be clear I agree that this was not the right thing to do.

**Comment by glozow** [comment]
Thanks for writing up a full review @ajtowns. Will keep this feedback in mind for the future if/when we get to this bridge again.

> implement cluster mempool first and export mempool.dat in chunks from highest to lowest fee; when importing, assume that the mempool contains chunks from highest score to lowest, build up a chunk to import; once it's over the mempool minfee, import the txs immediately; if it's over 1MvB and still under the mempool minfee, drop it and stop attempting to import txs from this mempool.dat entirely?

Yeah I'm pretty optimistic that cluster mempool will provide a natural solution like this.

**Comment by ajtowns** [comment]
> Yeah I'm pretty optimistic that cluster mempool will provide a natural solution like this.

Probably, but I suspect ephemeral anchors or similar will be ready before we've got all the i's dotted and t's crossed on cluster mempool, so some stop gap patch will be worthwhile. Still, we can cross that bridge once we get there.

**Comment by instagibbs** [comment]
Special-casing it is a value-add imo: v3/ephemeral anchor tx should be aggressively fee bumping anyways.

File: src/kernel/mempool_persist.cpp (hunk 1 above)
Comment (ajtowns) [reply src/kernel/mempool_persist.cpp:66]: As a datapoint, importing my mempool now (600MB used, 237k entries) seems to take about 25 minutes. Not a particularly fast machine (though not an rpi either), debug build, etc, so something of a worst-case scenario, but still.

**Pull request closed without merging**